enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,693 @@
+
+- `openshift-ingress` with `Route` resource management: non-operational without `CCM`.
+- `kubelet`: can't initialize a new `Node` without `CCM`.
+Consumes `cloudprovider.Interface` in order to collect `Node` [address](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L65), [populate](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L388-L435) `spec.ProviderID` and `Labels`,
+and [collect](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L960-L978) `Node` name.
+
+*Note: `storage`: in-tree plugins including storage implementation will remain untouched by the migration. This controller will remain preserved by using the `--external-cloud-volume-plugin` flag on `kcm` pod.*
+*Although `CCM` controller loops itself does not require HA setup as they does not store any state, the in-tree storage plugins are, and they will be removed together with the `CCM` from in-tree when it happens.*
+
+*The ability to support out-of-tree provider storage in a form of a CSI plugin and perform migration from in-tree storage plugin to the CSI one is the next step. Migration process design to CSI is described [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md). This is task is not a scope of this proposal.*
+
+#### Credentials management
+
+The operator is expected to be integrated with the [cloud-credentials-operator](https://github.com/openshift/cloud-credential-operator), and request fine grained credentials for the cloud components. It will be `CCO`'s responsibility to decide, depending on the platform, which credentials to request.
+
+*Bootstrap phase:*
+
+- Initial configuration with static pods is expected to be using `ConfigMap` credentials, similar to what is [done](https://github.com/openshift/installer/blob/ba6d7fe087eada5a4fc260064c85a484a8c45aaf/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L162) in the `kcm` operator.
+- Once it is supported, the static pods will also use the `CredentialsRequest` to get cloud credentials.
+
+*Post-install phase:*
+
+- Using a `CredentialsRequest` is the default option, every supported `CCM` within the `Deployment` will request its own set of credentials.
+- A set of `CredentialsRequest` resources will be hosted under `CCCMO` repository, and created by the `CVO`, similarly to the [machine-api](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml) approach.
+
+Depending on the cloud-provider, when it is supported, credentials management is deligated to the instance metadata endpoint to be acquired from. `AWS` and `Azure` are currently known to support this feature in some extent.
+
+#### Resource changes
+
+#### Kubelet
+
+In OpenShift, `machine-config-operator` manages `kubelet` configuration.
```

Comment:
Some verbiage regarding `windows-machine-config-operator` managing `kubelet` configuration for Windows nodes should be added here.

Reply:
Added a note that `kubelet` behaviour changes and effects would apply for `WMCO` as well.